pkg/controllers/utils/common.go

```diff
@@ -0,0 +1,117 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/karmada-io/karmada/pkg/util/helper"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+)
+
+func restrictFailoverHistoryInfo(binding *workv1alpha2.ResourceBinding) bool {
+	placement := binding.Spec.Placement
+	// Check if replica scheduling type is Duplicated
+	if placement.ReplicaScheduling.ReplicaSchedulingType == policyv1alpha1.ReplicaSchedulingTypeDuplicated {
+		return true
+	}
+
+	// Check if replica scheduling type is Divided with no spread constraints or invalid spread constraints
+	if placement.ReplicaScheduling.ReplicaSchedulingType == policyv1alpha1.ReplicaSchedulingTypeDivided {
+		if len(placement.SpreadConstraints) == 0 {
+			return true
+		}
+
+		for _, spreadConstraint := range placement.SpreadConstraints {
+			if spreadConstraint.SpreadByLabel != "" {
+				return true
+			}
+			if spreadConstraint.SpreadByField == "cluster" && (spreadConstraint.MaxGroups > 1 || spreadConstraint.MinGroups > 1) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func UpdateFailoverStatus(client client.Client, binding *workv1alpha2.ResourceBinding, cluster string, failoverType string) (err error) {
+	if restrictFailoverHistoryInfo(binding) {
+		return nil
+	}
+	message := fmt.Sprintf("Failover triggered for replica on cluster %s", cluster)
+
+	var reason string
+	if failoverType == workv1alpha2.EvictionReasonApplicationFailure {
+		reason = "ApplicationFailover"
+	} else if failoverType == workv1alpha2.EvictionReasonTaintUntolerated {
+		reason = "ClusterFailover"
+	} else {
+		errMsg := "Invalid failover type passed into updateFailoverStatus"
+		klog.Errorf(errMsg)
+		return fmt.Errorf(errMsg)
+	}
+
+	newFailoverAppliedCondition := metav1.Condition{
+		Type:               failoverType,
+		Status:             metav1.ConditionTrue,
+		Reason:             reason,
+		Message:            message,
+		LastTransitionTime: metav1.Now(),
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		_, err = helper.UpdateStatus(context.Background(), client, binding, func() error {
+			// set binding status with the newest condition
+			currentTime := metav1.Now()
+			failoverHistoryItem := workv1alpha2.FailoverHistoryItem{
+				FailoverTime:  &currentTime,
+				OriginCluster: cluster,
+				Reason:        reason,
+			}
+			binding.Status.FailoverHistory = append(binding.Status.FailoverHistory, failoverHistoryItem)
+			klog.V(4).Infof("Failover history is %+v", binding.Status.FailoverHistory)
+			existingCondition := meta.FindStatusCondition(binding.Status.Conditions, failoverType)
```

Comment:
Conditions are redundant, we can infer everything from the failover history. Will not populate conditions with failover information.